src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java

```diff
@@ -0,0 +1,950 @@
+                    ", tombstones=[" + String.join(";", tombstoneGens) + "]" +
+                    ", data=[" + String.join(";", dataGens) + "]" +
+                    "}";
+        }
+    }
+
+    private boolean isTooMuchGarbage(double totalSizeOnDiskBytes, double garbageBytes, boolean global)
+    {
+        double percentage = global ? options.windowCompactionGlobalSizePercent : options.windowCompactionSizePercent;
+        double absolute = global ? options.windowCompactionGlobalSizeInMB : options.windowCompactionSizeInMB;
+        double threshold = Double.min(totalSizeOnDiskBytes * (percentage / 100.0), absolute * FileUtils.ONE_MB);
+        return garbageBytes >= threshold;
+    }
+
+    private static class TombstoneCounts
+    {
+        long partitionTombstones = 0;
+        long rowTombstones = 0;
+        long rangeTombstones = 0;
+
+        public void add(TombstoneCounts counts)
+        {
+            partitionTombstones += counts.partitionTombstones;
+            rowTombstones += counts.rowTombstones;
+            rangeTombstones += counts.rangeTombstones;
+        }
+
+        @Override
+        public String toString()
+        {
+            return "TombstoneCounts{" +
+                    "p=" + partitionTombstones +
+                    ", r=" + rowTombstones +
+                    ", rng=" + rangeTombstones +
+                    '}';
+        }
+    }
+
+    static class CategorizedSSTables
+    {
+        public final List<SSTableStats> data;
+        public final List<SSTableStats> expiredTombstones;
+        public final List<SSTableStats> nearExpiryTombstones;
+        public final List<SSTableStats> latestTombstones;
+        public final OverallStats stats;
+
+        public CategorizedSSTables(List<SSTableStats> data, List<SSTableStats> expiredTombstones, List<SSTableStats> nearExpiryTombstones, List<SSTableStats> latestTombstones, OverallStats stats)
+        {
+            this.data = data;
+            this.expiredTombstones = expiredTombstones;
+            this.nearExpiryTombstones = nearExpiryTombstones;
+            this.latestTombstones = latestTombstones;
+            this.stats = stats;
+        }
+
+        @Override
+        public String toString()
+        {
+            return "CategorizedSSTables{" +
+                    "data=" + idString(data) +
+                    ", expiredTombstones=" + idString(expiredTombstones) +
+                    ", nearExpiryTombstones=" + idString(nearExpiryTombstones) +
+                    ", latestTombstones=" + idString(latestTombstones) +
+                    ", stats=" + stats +
+                    '}';
+        }
+    }
+
+    CategorizedSSTables categorizeSStables(Iterable<SSTableReader> sstables, int gcBefore)
+    {
+        int gcGraceSecondsMs = Math.max(1, cfs.metadata.params.gcGraceSeconds) * 1000;
+        long nowMs = System.currentTimeMillis();
+        long currentGcWindowMs = toWindowMs(nowMs, gcGraceSecondsMs);
```

Comment:
Is the window size for compaction always same as gc grace seconds?

Reply:
yes, gc grace seconds can be zero. 
Math.max(1, cfs.metadata.params.gcGraceSeconds) is kinda useless now. As when i am using the variable "gcGraceSecondsMs", i am explicitly checking for gcgraceSeconds to be nonZero.